{
 "metadata": {
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9-final"
  },
  "orig_nbformat": 2,
  "kernelspec": {
   "name": "python36964bit96b38543bc524d45aa4022ceee87892b",
   "display_name": "Python 3.6.9 64-bit",
   "language": "python"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2,
 "cells": [
  {
   "source": [
    "## Find a Position in a Genome Minimizing the Skew \n"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def skew(genome):\n",
    "    val = 0;\n",
    "    track = []\n",
    "    track.append(0)\n",
    "    for i in genome:\n",
    "        if i == \"G\":\n",
    "            val += 1\n",
    "        elif i == \"C\":\n",
    "            val -=1\n",
    "        track.append(val)\n",
    "\n",
    "    # for k in track:\n",
    "    #     print(k, end=\" \")\n",
    "    return track\n",
    "skew(\"GAGCCACCGCGATA\")\n",
    "def skew_minima(genome):\n",
    "    track = skew(genome)\n",
    "    idx_list = []\n",
    "    # print(track)\n",
    "    minima = min(track)\n",
    "    for i in range(len(track)):\n",
    "        if (track[i] == minima):\n",
    "            idx_list.append(i)\n",
    "    for i in idx_list:\n",
    "        print(i,end=\" \")\n",
    "skew_minima(\"\")"
   ]
  },
  {
   "source": [
    "## Find Patterns Forming Clumps in a String"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def NumberToPattern(index, k):\n",
    "    table = {0:'A', 1:'C', 2:'G', 3:'T'}\n",
    "    if k == 1:\n",
    "        # print(\"ret for k: \",k ,\" index : \",int(index) , \" => \" ,table[int(index)])\n",
    "        return table[int(index)]\n",
    "    prefixIndex = index / 4\n",
    "    r = index % 4\n",
    "    prefixPattern = NumberToPattern(prefixIndex, k-1)\n",
    "    symbol = table[int(r)]\n",
    "    # print(\"ret for k: \",k ,\" index : \",int(index) , \" => \" , prefixPattern + symbol)\n",
    "    return prefixPattern + symbol\n",
    "\n",
    "def SymbolToNumber(symbol):\n",
    "    if symbol=='A':\n",
    "        return 0\n",
    "    elif symbol=='C':\n",
    "        return 1\n",
    "    elif symbol=='G':\n",
    "        return 2\n",
    "    else:\n",
    "        return 3\n",
    "    \n",
    "def PatternToNumber(pattern):\n",
    "    if pattern=='':\n",
    "        return 0\n",
    "    symbol=pattern[-1]\n",
    "    prefix=pattern[:-1]\n",
    "    return 4*PatternToNumber(prefix)+SymbolToNumber(symbol)\n",
    "\n",
    "def FrequenceArray(sequence, k):\n",
    "    FrequencyArray = []\n",
    "    for i in range(4**k):\n",
    "        FrequencyArray.append(0)\n",
    "    for i in range(len(sequence) - k + 1):\n",
    "        kmer = sequence[i:i+k]\n",
    "        index = PatternToNumber(kmer) #This is for hashing the kmer sequence\n",
    "        # print(\"FA : \" , FrequencyArray[index])\n",
    "        FrequencyArray[index] = FrequencyArray[index] + 1\n",
    "    return FrequencyArray\n",
    "\n",
    "def ClumpFinding(Genome, k, t, L):\n",
    "    clumps = []\n",
    "    FrequencePattern = set()\n",
    "    for i in range(4**k - 1):\n",
    "        clumps.append(0)\n",
    "    for i in range(len(Genome) - L):\n",
    "        window = Genome[i:i+L]\n",
    "        FrequenceArray = FrequenceArray(window, k)\n",
    "        for j in range(4**k - 1):\n",
    "            if FrequenceArray[j] >= t:\n",
    "                clumps[j] = 1\n",
    "    for i in range(4**k - 1):\n",
    "        if clumps[i] == 1:\n",
    "            pattern = NumberToPattern(i, k)\n",
    "            FrequencePattern.add(pattern)\n",
    "    return FrequencePattern"
   ]
  },
  {
   "source": [
    "## Find a Median String "
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def HammingDistance(p, q):\n",
    "    count = 0\n",
    "    for (i,j) in zip(p,q):\n",
    "        if i != j:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "def getKmers(dna, k):\n",
    "    for i in range(len(dna) - k + 1):\n",
    "        yield dna[i:i+k]\n",
    "\n",
    "def DiffBetweenPatternandStrings(pattern, dna): # minimizes the sum of hamming distances to the given dnas with a pattern\n",
    "    k =  len(pattern)\n",
    "    diff = 0;\n",
    "    for d in dna:\n",
    "        ham_distance = 2 ** 31\n",
    "        for i in range (len(d) - k + 1):\n",
    "            pattern_ = d[i : i+k]\n",
    "            if (HammingDistance(pattern,pattern_) < ham_distance):\n",
    "                ham_distance = HammingDistance(pattern,pattern_)\n",
    "        diff = diff + ham_distance\n",
    "\n",
    "    return diff\n",
    "\n",
    "def MedianString(dna, k) : # A string that minimizes the sum of distances to the strings of a given set \n",
    "    diff = 2 ** 31 \n",
    "    kmers = generate_strings(k)\n",
    "    median = \"\"\n",
    "    for kmer in kmers:\n",
    "        temp_diff =  DiffBetweenPatternandStrings(kmer, dna)\n",
    "        if (temp_diff < diff) :\n",
    "            diff =  temp_diff\n",
    "            median = kmer\n",
    "\n",
    "    return median"
   ]
  },
  {
   "source": [
    "## Find a Profile-most Probable k-mer in a String"
   ],
   "cell_type": "markdown",
   "metadata": {}
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculateProbability(pattern, profile_mat): # given a profile, calculates the probability of a string occuring \n",
    "    temp_pr = 1.0\n",
    "    for j in range(len(pattern)):\n",
    "        if(pattern[j] == \"A\"):\n",
    "            temp_pr  = temp_pr * profile_mat [\"A\"][j]\n",
    "        elif(pattern[j] == \"C\"):\n",
    "            temp_pr  = temp_pr * profile_mat [\"C\"][j]\n",
    "        elif(pattern[j] == \"G\"):\n",
    "            temp_pr  = temp_pr * profile_mat [\"G\"][j]\n",
    "        elif(pattern[j] == \"T\"):\n",
    "            temp_pr  = temp_pr * profile_mat [\"T\"][j]\n",
    "    return temp_pr\n",
    "\n",
    "\n",
    "def ProfileMostProbableKmer(text, k, profile):\n",
    "    max_p = -1\n",
    "    most_probable_kmer = ''\n",
    "    for i in range(len(text)-k+1):\n",
    "        pattern = text[i: i+k]\n",
    "        p = calculateProbability(pattern, profile)\n",
    "        if p > max_p:\n",
    "            max_p = p\n",
    "            most_probable_kmer = pattern           \n",
    "    return most_probable_kmer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def Count(Motifs):\n",
    "    count = {} # initializing the 4xk count dictionary\n",
    "    k = len(Motifs[0])\n",
    "    count = {'A':[0]*k,'C':[0]*k,'G':[0]*k,'T':[0]*k}\n",
    "    \n",
    "    t = len(Motifs)\n",
    "    for i in range(t):\n",
    "        for j in range(k):\n",
    "            symbol = Motifs[i][j]\n",
    "            count[symbol][j] += 1\n",
    "    return count\n",
    "\n",
    "def Consensus(Motifs): # get the consensus string for given motifs by picking the most frequent symbol in each column\n",
    "    k = len(Motifs[0])\n",
    "    count = Count(Motifs)\n",
    "    consensus = \"\"\n",
    "    for j in range(k):\n",
    "        m = 0\n",
    "        frequentSymbol = \"\"\n",
    "        for symbol in \"ACGT\":\n",
    "            if count[symbol][j] > m:\n",
    "                m = count[symbol][j]\n",
    "                frequentSymbol = symbol\n",
    "        consensus += frequentSymbol\n",
    "    return consensus\n",
    "\n",
    "def Profile(Motifs): ## generate profile matrix from motifs\n",
    "    t = len(Motifs)\n",
    "    k = len(Motifs[0])\n",
    "    profile = Count(Motifs)\n",
    "    print(\"Count: \\n\", profile)\n",
    "\n",
    "    for i in profile:  \n",
    "        for j in range(k):\n",
    "           profile[i][j] = float(profile[i][j])/float(t)   \n",
    "    return profile\n",
    "\n",
    "def Score(Motifs):\n",
    "    consensus = Consensus(Motifs) #consensus string\n",
    "    counts = Count(Motifs) # count matrix\n",
    "    score = len(Motifs)*len(consensus) \n",
    "    i = 0\n",
    "    for symbol in consensus:  \n",
    "        score -= counts[symbol][i] \n",
    "        i += 1\n",
    "    return score \n",
    "\n",
    "def GreedyMotifSearch(Dna, k, t):\n",
    "    BestMotifs = []\n",
    "    for i in range(0, t):\n",
    "        BestMotifs.append(Dna[i][0:k]) # first k-mers from each of the dna strings\n",
    "    \n",
    "    n = len(Dna[0]) \n",
    "    for i in range(n-k+1):\n",
    "        Motifs = []\n",
    "        Motifs.append(Dna[0][i:i+k]) # kmer from 1st string\n",
    "        for j in range(1, t): #for rest of the strings\n",
    "            P = Profile(Motifs[0:j]) # successively calculates profile matrix by adding one string per iteration\n",
    "            print(\"intmed Profile for \", Motifs[0:j])\n",
    "            print(P)\n",
    "            Motifs.append(ProfileMostProbableKmer(Dna[j], k, P))\n",
    "            print(\"PMP kmer for \",Dna[j],\" is: \",ProfileMostProbableKmer(Dna[j], k, P))\n",
    "\n",
    "        print(\"current motifs: \", Motifs)\n",
    "        if Score(Motifs) < Score(BestMotifs):\n",
    "            BestMotifs = Motifs\n",
    "        print(\"..........................................................................................\")\n",
    "    return BestMotifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "output_type": "stream",
     "name": "stdout",
     "text": [
      "Count: \n {'A': [1, 0, 0, 0], 'C': [0, 1, 1, 0], 'G': [0, 0, 0, 0], 'T': [0, 0, 0, 1]}\nintmed Profile for  ['ACCT']\n{'A': [1.0, 0.0, 0.0, 0.0], 'C': [0.0, 1.0, 1.0, 0.0], 'G': [0.0, 0.0, 0.0, 0.0], 'T': [0.0, 0.0, 0.0, 1.0]}\nPMP kmer for  AGTTGG  is:  AGTT\nCount: \n {'A': [2, 0, 0, 0], 'C': [0, 1, 1, 0], 'G': [0, 1, 0, 0], 'T': [0, 0, 1, 2]}\nintmed Profile for  ['ACCT', 'AGTT']\n{'A': [1.0, 0.0, 0.0, 0.0], 'C': [0.0, 0.5, 0.5, 0.0], 'G': [0.0, 0.5, 0.0, 0.0], 'T': [0.0, 0.0, 0.5, 1.0]}\nPMP kmer for  CCGTTC  is:  CCGT\nCount: \n {'A': [2, 0, 0, 0], 'C': [1, 2, 1, 0], 'G': [0, 1, 1, 0], 'T': [0, 0, 1, 3]}\nintmed Profile for  ['ACCT', 'AGTT', 'CCGT']\n{'A': [0.6666666666666666, 0.0, 0.0, 0.0], 'C': [0.3333333333333333, 0.6666666666666666, 0.3333333333333333, 0.0], 'G': [0.0, 0.3333333333333333, 0.3333333333333333, 0.0], 'T': [0.0, 0.0, 0.3333333333333333, 1.0]}\nPMP kmer for  ACGTAA  is:  ACGT\ncurrent motifs:  ['ACCT', 'AGTT', 'CCGT', 'ACGT']\n..........................................................................................\nCount: \n {'A': [0, 0, 0, 1], 'C': [1, 1, 0, 0], 'G': [0, 0, 0, 0], 'T': [0, 0, 1, 0]}\nintmed Profile for  ['CCTA']\n{'A': [0.0, 0.0, 0.0, 1.0], 'C': [1.0, 1.0, 0.0, 0.0], 'G': [0.0, 0.0, 0.0, 0.0], 'T': [0.0, 0.0, 1.0, 0.0]}\nPMP kmer for  AGTTGG  is:  AGTT\nCount: \n {'A': [1, 0, 0, 1], 'C': [1, 1, 0, 0], 'G': [0, 1, 0, 0], 'T': [0, 0, 2, 1]}\nintmed Profile for  ['CCTA', 'AGTT']\n{'A': [0.5, 0.0, 0.0, 0.5], 'C': [0.5, 0.5, 0.0, 0.0], 'G': [0.0, 0.5, 0.0, 0.0], 'T': [0.0, 0.0, 1.0, 0.5]}\nPMP kmer for  CCGTTC  is:  CGTT\nCount: \n {'A': [1, 0, 0, 1], 'C': [2, 1, 0, 0], 'G': [0, 2, 0, 0], 'T': [0, 0, 3, 2]}\nintmed Profile for  ['CCTA', 'AGTT', 'CGTT']\n{'A': [0.3333333333333333, 0.0, 0.0, 0.3333333333333333], 'C': [0.6666666666666666, 0.3333333333333333, 0.0, 0.0], 'G': [0.0, 0.6666666666666666, 0.0, 0.0], 'T': [0.0, 0.0, 1.0, 0.6666666666666666]}\nPMP kmer for  ACGTAA  is:  CGTA\ncurrent motifs:  ['CCTA', 'AGTT', 'CGTT', 'CGTA']\n..........................................................................................\nCount: \n {'A': [0, 0, 1, 0], 'C': [1, 0, 0, 0], 'G': [0, 0, 0, 0], 'T': [0, 1, 0, 1]}\nintmed Profile for  ['CTAT']\n{'A': [0.0, 0.0, 1.0, 0.0], 'C': [1.0, 0.0, 0.0, 0.0], 'G': [0.0, 0.0, 0.0, 0.0], 'T': [0.0, 1.0, 0.0, 1.0]}\nPMP kmer for  AGTTGG  is:  AGTT\nCount: \n {'A': [1, 0, 1, 0], 'C': [1, 0, 0, 0], 'G': [0, 1, 0, 0], 'T': [0, 1, 1, 2]}\nintmed Profile for  ['CTAT', 'AGTT']\n{'A': [0.5, 0.0, 0.5, 0.0], 'C': [0.5, 0.0, 0.0, 0.0], 'G': [0.0, 0.5, 0.0, 0.0], 'T': [0.0, 0.5, 0.5, 1.0]}\nPMP kmer for  CCGTTC  is:  CGTT\nCount: \n {'A': [1, 0, 1, 0], 'C': [2, 0, 0, 0], 'G': [0, 2, 0, 0], 'T': [0, 1, 2, 3]}\nintmed Profile for  ['CTAT', 'AGTT', 'CGTT']\n{'A': [0.3333333333333333, 0.0, 0.3333333333333333, 0.0], 'C': [0.6666666666666666, 0.0, 0.0, 0.0], 'G': [0.0, 0.6666666666666666, 0.0, 0.0], 'T': [0.0, 0.3333333333333333, 0.6666666666666666, 1.0]}\nPMP kmer for  ACGTAA  is:  ACGT\ncurrent motifs:  ['CTAT', 'AGTT', 'CGTT', 'ACGT']\n..........................................................................................\n"
     ]
    },
    {
     "output_type": "execute_result",
     "data": {
      "text/plain": [
       "['ACCT', 'AGTT', 'CCGT', 'ACGT']"
      ]
     },
     "metadata": {},
     "execution_count": 33
    }
   ],
   "source": [
    "GreedyMotifSearch([\"ACCTAT\", \"AGTTGG\", \"CCGTTC\", \"ACGTAA\"], 4, 4)"
   ]
  }
 ]
}